cmd/grpcurl: add -allow-unknown-fields option

adambabik opened the pull request:
This PR adds a new option `-allow-unknown-fields` (its default value is `false` so the original behavior does not change) which, if present, will instruct JSON protobuf parser to ignore unknown fields present in the request.

My use case is that I have two services sharing a message. The message got extend with a new field but the services haven't been upgraded yet. The change to the message is backward compatible but I can't use `grpcurl` due to an error: `error getting request data: message type ABC has no known field named my_new_field`.

adambabik commented:
@jhump thanks for checking. You're totally right, in this form the package itself requires a major version bump. I will reformulate this PR to avoid that.

adambabik commented:
@jhump sorry, it took me actually a while to find some spare time. I rewrote the code so that it does not alter the public interface.

jhump commented on format.go:
```diff
@@ -342,23 +351,46 @@ func (a *unknownAny) ProtoMessage() {
+type FormatterOptions struct {
+	EmitJSONDefaultFields bool // FormatJSON only
+	IncludeTextSeparator  bool // FormatText only
+	AllowUnknownFields    bool // FormatJSON only
```
Let's group the JSON related fields together, instead of interleaving.

jhump replied:
Technically `AllowUnknownFields` is an option for the parser, not the formatter.

How about rename this to `FormatOptions` -- since they can apply to parsing the format, not just outputting it.

jhump commented on format.go:
```diff
@@ -342,23 +351,46 @@ func (a *unknownAny) ProtoMessage() {
+// are options for JSON and protobuf text formats, respectively. The AllowUnknownFields field
+// is a JSON-only format flag.
+// Requests will be parsed from the given in.
+func RequestParserAndFormatterWithOptions(format Format, descSource DescriptorSource, opts FormatterOptions, in io.Reader) (RequestParser, Formatter, error) {
```
Let's shorten this to `RequestParserAndFormatter` and then deprecate the other version.

Also, please move `opts` to be the last argument since that is more idiomatic for functions that use an options-struct pattern.

jhump commented on format.go:
```diff
@@ -342,23 +351,46 @@ func (a *unknownAny) ProtoMessage() {
+// includeTextSeparator are options for JSON and protobuf text formats,
+// respectively. Requests will be parsed from the given in.
+func RequestParserAndFormatterFor(format Format, descSource DescriptorSource, emitJSONDefaultFields, includeTextSeparator bool, in io.Reader) (RequestParser, Formatter, error) {
+	opts := FormatterOptions{
```
After moving `opts` to end of arg list for subsequent call, this could be inlined.

jhump commented on cmd/grpcurl/grpcurl.go:
```diff
@@ -615,7 +619,12 @@ func main() {
+				options := grpcurl.FormatterOptions{
+					EmitJSONDefaultFields: true,
+					IncludeTextSeparator:  false,
+					AllowUnknownFields:    *allowUnknownFields,
```
Let's just hardcode to `false`, since this code path won't be parsing anything.

Also, you can just omit the literal false values since they are default/zero vals:

```
options := grpcurl.FormatOptions{EmitJSONDefaultFields: true}
```

adambabik commented on format.go:
```diff
@@ -342,23 +351,60 @@ func (a *unknownAny) ProtoMessage() {
+// includeTextSeparator are options for JSON and protobuf text formats,
+// respectively. Requests will be parsed from the given in.
+// This function is deprecated. Please use RequestParserAndFormatter instead.
+// DEPRECATED
```
I am not sure what's project's convention to mark functions as deprecated. I explained it in the comment and added `// DEPRECATED` line. The latter works nicely with editors, for example GoLand, that inform when a deprecated function is used.

jhump replied:
https://github.com/golang/go/wiki/Deprecated

This is fine. I may tweak it in a later commit. More idiomatic for Go would be to take the last two lines here and change them to something like so:

```go
// Deprecated: Use RequestParserAndFormatter instead.
```

adambabik commented:
@jhump good stuff, thanks for a detailed review. I applied all suggestions, added a bit more code comments, and fixed tests to use `RequestParserAndFormatter` instead of now deprecated `RequestParserAndFormatterFor`.
